src/idpf.rs

```diff
@@ -0,0 +1,1852 @@
+    start_level: usize,
+    mut seed: [u8; L],
+    mut control_bit: Choice,
+    prefix: &IdpfInput,
+    cache: &mut dyn IdpfCache<L>,
+) -> Result<IdpfOutputShare<OUT_LEN, Field64, Field255>, IdpfError>
+where
+    P: Prg<L>,
+{
+    let mut last_inner_output = None;
+    for ((correction_word, input_bit), level) in public_share.inner_correction_words[start_level..]
+        .iter()
+        .zip(prefix[start_level..].iter())
+        .zip(start_level..)
+    {
+        last_inner_output = Some(eval_next::<_, P, L, OUT_LEN>(
+            which_aggregator,
+            &mut seed,
+            &mut control_bit,
+            correction_word,
+            Choice::from(*input_bit as u8),
+        ));
+        let cache_key = &prefix[..=level];
+        cache.insert(cache_key, &(seed, control_bit.unwrap_u8()));
+    }
+
+    if prefix.len() == bits {
+        let leaf_output = eval_next::<_, P, L, OUT_LEN>(
+            which_aggregator,
+            &mut seed,
+            &mut control_bit,
+            &public_share.leaf_correction_word,
+            Choice::from(prefix[bits - 1] as u8),
+        );
+        // Note: there's no point caching this seed, because we will always run the eval_next()
+        // call for the leaf level.
+        Ok(IdpfOutputShare::Leaf(leaf_output))
+    } else {
+        Ok(IdpfOutputShare::Inner(last_inner_output.unwrap()))
+    }
+}
+
+/// The IdpfPoplar key evaluation algorithm.
+///
+/// Evaluate an IDPF share on `prefix`.
+pub fn eval<P, const L: usize, const OUT_LEN: usize>(
+    bits: usize,
+    agg_id: usize,
+    public_share: &IdpfPoplarPublicShare<Field64, Field255, L, OUT_LEN>,
+    key: &Seed<L>,
+    prefix: &IdpfInput,
+    cache: &mut dyn IdpfCache<L>,
+) -> Result<IdpfOutputShare<OUT_LEN, Field64, Field255>, IdpfError>
+where
+    P: Prg<L>,
+{
+    if bits == 0 {
+        return Err(IdpfError::InvalidParameter(
+            "invalid number of bits: 0".to_string(),
+        ));
+    }
+    if prefix.is_empty() {
+        return Err(IdpfError::InvalidParameter("empty prefix".to_string()));
+    }
+    if prefix.len() > bits {
+        return Err(IdpfError::InvalidParameter(format!(
+            "prefix length ({}) exceeds configured number of bits ({})",
+            prefix.len(),
+            bits,
+        )));
+    }
+
+    if public_share.inner_correction_words.len() != bits - 1 {
+        return Err(IdpfError::InvalidParameter(format!(
+            "public share had wrong number of correction words ({}, expected {})",
+            public_share.inner_correction_words.len() + 1,
+            bits,
+        )));
+    }
+
+    let which_aggregator = agg_id != 0;
+
+    // Check for cached seeds first.
+    if prefix.len() > 1 {
+        // Skip checking for `prefix` in the cache, because we don't store field element
+        // values along with seeds and control bits. Instead, start looking one node higher
+        // up, so we can recompute everything for the last level of `prefix`.
```

Comment:
Actual evaluation of the IDPF does start with level zero and `prefix[0]`, but this loop is intentionally operating in the reverse order because it's looking for the node at the greatest level where it can get a cache hit. I'll expand on the comments around this.

Reply:
Phew, the comments clear this up. Thanks for clarifying!